File: android/app/src/main/kotlin/com/vyooo/insta360/pipeline/PipelineHints.kt
package com.vyooo.insta360.pipeline

/**
 * Optional decision signals for the pipeline — metadata only, no pixel data. An on-device AI layer
 * may supply these; the pipeline runs fully when a hint is `null` (each field is nullable and stages
 * treat `null` as "use the deterministic default").
 */
interface PipelineHints {
    /** Stabilised forward direction θ₀ (degrees); `null` → use the frame default. */
    val forwardThetaDeg: Float?

    /** Classification override for panoramic-vs-planar; `null` → let the detector decide. */
    val isPanoramic: Boolean?

    /** Perceptual-salience hint guiding the downscale target; `null` → fixed target. */
    val perceptualScale: Float?

    /** Motion metric in [0,1] for temporal gating; `null` → the stage computes its own. */
    val motion: Float?
}

/** Defaults with no AI: every hint absent, so stages use deterministic behaviour. */
object DeterministicHints : PipelineHints {
    override val forwardThetaDeg: Float? = null
    override val isPanoramic: Boolean? = null
    override val perceptualScale: Float? = null
    override val motion: Float? = null
}

/**
 * Live-updatable hints written by an on-device AI layer and read by the pipeline once per frame.
 * Only what the AI sets is applied; everything left `null` falls back to deterministic logic.
 */
class MutableHints : PipelineHints {
    @Volatile override var forwardThetaDeg: Float? = null
    @Volatile override var isPanoramic: Boolean? = null
    @Volatile override var perceptualScale: Float? = null
    @Volatile override var motion: Float? = null
}

File: ios/Runner/Insta360/Pipeline/PipelineHints.swift
import Foundation

/// Optional decision signals for the pipeline — metadata only, no pixel data. An on-device AI layer
/// may supply these; the pipeline runs fully when a hint is `nil`. Swift port of Android
/// `PipelineHints.kt` (interface + `DeterministicHints` + `MutableHints`).
protocol PipelineHints: AnyObject {
    /// Stabilised forward direction θ₀ (degrees); `nil` → use the frame default.
    var forwardThetaDeg: Float? { get }
    /// Classification override for panoramic-vs-planar; `nil` → let the detector decide.
    var isPanoramic: Bool? { get }
    /// Perceptual-salience hint guiding the downscale target; `nil` → fixed target.
    var perceptualScale: Float? { get }
    /// Motion metric in [0,1] for temporal gating; `nil` → the stage computes its own.
    var motion: Float? { get }
}

/// Defaults with no AI: every hint absent, so stages use deterministic behaviour.
final class DeterministicHints: PipelineHints {
    static let shared = DeterministicHints()
    var forwardThetaDeg: Float? { nil }
    var isPanoramic: Bool? { nil }
    var perceptualScale: Float? { nil }
    var motion: Float? { nil }
}

/// Live-updatable hints written by the on-device AI layer and read by the pipeline once per frame.
/// Only what the AI sets is applied; everything left `nil` falls back to deterministic logic.
final class MutableHints: PipelineHints {
    var forwardThetaDeg: Float?
    var isPanoramic: Bool?
    var perceptualScale: Float?
    var motion: Float?
}
